contracts/modules/TransferManager/VolumeRestrictionTM.sol

```diff
@@ -0,0 +1,934 @@
+                );
+            }
+            return Result.NA;
+        }
+        return Result.INVALID;
+    }
+
+    /**
+     * @notice Internal function to update the state variables related to global restriction
+     */
+    function _updateGlobalStorage(
+        uint256[] passedTimestamps,
+        uint256 _fromTime,
+        uint256 _amount,
+        uint256 _diffDays,
+        uint256 _sumOfLastPeriod
+    )
+        internal
+    {
+        if (_diffDays != 0) {
+            for (uint256 i = 0; i < passedTimestamps.length; i++) {
+                // Add the timestamp that is already passed
+                globalBucketDetails.timestamps.push(passedTimestamps[i]);
+            }
+        }
+        // This condition is the works only when the transaction performed just after the startTime (_diffDays == 0)
+        if (globalBucketDetails.timestamps.length == 0) {
+            globalBucketDetails.timestamps.push(_fromTime);
+        }
+        if(_amount != 0) {
+            // updating the sumOfLastPeriod
+            globalBucketDetails.sumOfLastPeriod = _sumOfLastPeriod.add(_amount);
+
+            // Re-using the local variable to avoid stack too deep error
+            _fromTime = globalBucketDetails.timestamps[globalBucketDetails.timestamps.length -1];
+            globalBucket[_fromTime] = globalBucket[_fromTime].add(_amount);
+        }
+    }
+
+    /// Internal function for the bucket check
+    function _bucketCheck(
+        uint256 _fromTime,
+        uint256 _diffDays,
+        address _from,
+        BucketDetails _bucketDetails,
+        bool _isGlobal,
+        VolumeRestriction _restriction
+    )
+        internal
+        view
+        returns (uint256, uint256, uint256[])
+    {
+        uint256[] memory passedTimestamps = new uint256[](_diffDays);
+        uint256 counter = _bucketDetails.timestamps.length;
+        uint256 i = 0;
+        if (_diffDays != 0) {
+            for (i = 0; i < _diffDays; i++) {
+                // calculating the timestamp that will used as an index of the next bucket
+                // i.e buckets period will be look like this T1 to T2-1, T2 to T3-1 .... 
+                // where T1,T2,T3 are timestamps having 24 hrs difference
+                _fromTime = _fromTime.add(1 days);
+
+                // This condition is to check whether the first rolling period is covered or not
+                // if not then it continues and adding 0 value into sumOfLastPeriod without subtracting
+                // the earlier value at that index
+                if (counter >= _restriction.rollingPeriodInDays) {
+                    if (_isGlobal) {
+                        // Subtracting the former value(Sum of all the txn amount of that day) from the sumOfLastPeriod
+                        _bucketDetails.sumOfLastPeriod = _bucketDetails.sumOfLastPeriod.
+                        sub(globalBucket[_bucketDetails.timestamps[counter.sub(_restriction.rollingPeriodInDays)]]);
```

Comment:
This will give the `invalid opcode` error when `_bucketDetails.timestamps` does not have enough timestamps pushed. It can happen as global storage is updated at the end and hence the array might be empty when starting out the calculations.

"Index out of bounds" error in normal terms.